File: InfixToPostfix.py
class CharStack:
    def __init__(self, size):
        self.data = [''] * size
        self.top = -1

    def push(self, c):
        self.top += 1
        self.data[self.top] = c

    def pop(self):
        val = self.data[self.top]
        self.top -= 1
        return val

    def peek(self):
        return self.data[self.top]

    def isEmpty(self):
        return self.top == -1


class IntStack:
    def __init__(self, size):
        self.data = [0] * size
        self.top = -1

    def push(self, x):
        self.top += 1
        self.data[self.top] = x

    def pop(self):
        val = self.data[self.top]
        self.top -= 1
        return val

    def isEmpty(self):
        return self.top == -1


def prec(c):
    if c == '^':
        return 3
    elif c == '/' or c == '*':
        return 2
    elif c == '+' or c == '-':
        return 1
    else:
        return -1


def isRightAssociative(c):
    return c == '^'


def infixToPostfix(s):
    st = CharStack(len(s))
    res = []

    i = 0
    while i < len(s):
        c = s[i]

        if c.isdigit():

            while i < len(s) and s[i].isdigit():
                res.append(s[i])
                i += 1

            res.append(' ')
            i -= 1

        elif c == '(':
            st.push('(')

        elif c == ')':
            while not st.isEmpty() and st.peek() != '(':
                res.append(st.pop())
            st.pop()

        else:
            while (not st.isEmpty() and st.peek() != '(' and
                   (prec(st.peek()) > prec(c) or
                   (prec(st.peek()) == prec(c) and not isRightAssociative(c)))):
                res.append(st.pop())
            st.push(c)

        i += 1

    while not st.isEmpty():
        res.append(st.pop())

    return ''.join(res)


def floorDiv(a, b):
    if a * b < 0 and a % b != 0:
        return (a // b) - 1
    return a // b


def evaluatePostfix(postfix):
    st = IntStack(len(postfix))

    i = 0
    while i < len(postfix):

        token = postfix[i]

        if token == ' ':
            i += 1
            continue

        if token.isdigit():

            num = 0

            while i < len(postfix) and postfix[i].isdigit():
                num = num * 10 + (ord(postfix[i]) - ord('0'))
                i += 1

            st.push(num)
            i -= 1

        else:
            val1 = st.pop()
            val2 = st.pop()

            result = 0

            if token == '+':
                result = val2 + val1
            elif token == '-':
                result = val2 - val1
            elif token == '*':
                result = val2 * val1
            elif token == '/':
                result = val2 // val1
            elif token == '^':
                result = int(val2 ** val1)

            print(val2, token, val1, "=", result)

            st.push(result)

        i += 1

    return st.pop()


# main program
exp = input("Masukkan ekspresi: ")
exp = exp.replace(" ", "")

postfix = infixToPostfix(exp)

print()
print("Postfix:", postfix)
print()

hasil = evaluatePostfix(postfix)

print()
print("Hasil akhir=", hasil)

File: InfixToPostfix.java
import java.util.Scanner;

public class InfixToPostfix {

 static class CharStack {

    private char[] data;
    private int top;

    CharStack(int size) {
        data = new char[size];
        top = -1;
    }

    void push(char c) {
        data[++top] = c;
    }

    char pop() {
        return data[top--];
    }

    char peek() {
        return data[top];
    }

    boolean isEmpty() {
        return top == -1;
    }
}

static class IntStack {

    private int[] data;
    private int top;

    IntStack(int size) {
        data = new int[size];
        top = -1;
    }

    void push(int x) {
        data[++top] = x;
    }

    int pop() {
        return data[top--];
    }

    boolean isEmpty() {
        return top == -1;
    }
}
    
    static int prec(char c) {
        if (c == '^')
            return 3;
        else if (c == '/' || c == '*')
            return 2;
        else if (c == '+' || c == '-')
            return 1;
        else
            return -1;
    }

    static boolean isRightAssociative(char c) {
        return c == '^';
    }

    public static String infixToPostfix(String s) {
        CharStack st = new CharStack(s.length());
        StringBuilder res = new StringBuilder();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (Character.isDigit(c)) {

            while (i < s.length() && Character.isDigit(s.charAt(i))) {
                res.append(s.charAt(i));
                i++;
            }

                res.append(' ');
                i--;
            }

            else if (c == '(')
                st.push('(');

            else if (c == ')') {
                while (!st.isEmpty() && st.peek() != '(') {
                    res.append(st.pop());
                }
                st.pop();
            }

            else {
                while (!st.isEmpty() && st.peek() != '(' &&
                       (prec(st.peek()) > prec(c) ||
                       (prec(st.peek()) == prec(c) && !isRightAssociative(c)))) {
                    res.append(st.pop());
                }
                st.push(c);
            }
        }

        while (!st.isEmpty()) {
            res.append(st.pop());
        }

        return res.toString();
    }


    static int floorDiv(int a, int b) {
        if (a * b < 0 && a % b != 0)
            return (a / b) - 1;
        return a / b;
    }
    
    public static int evaluatePostfix(String postfix) {

    IntStack st = new IntStack(postfix.length());

    for (int i = 0; i < postfix.length(); i++) {

        char token = postfix.charAt(i);

        if (token == ' ')
            continue;

        if (Character.isDigit(token)) {

    int num = 0;

    while (i < postfix.length() && Character.isDigit(postfix.charAt(i))) {
        num = num * 10 + (postfix.charAt(i) - '0');
        i++;
    }

        st.push(num);
        i--;
    }
        else {
            int val1 = st.pop();
            int val2 = st.pop();

            int result = 0;

            if (token == '+') result = val2 + val1;
            else if (token == '-') result = val2 - val1;
            else if (token == '*') result = val2 * val1;
            else if (token == '/') result = val2 / val1;
            else if (token == '^')
                result = (int)Math.pow(val2, val1);

            System.out.println(
                val2 + " " + token + " " + val1 + " = " + result
            );

            st.push(result);
        }
    }

    return st.pop();
}



public static void main(String[] args) {

    Scanner sc = new Scanner(System.in);

    System.out.print("Masukkan ekspresi: ");
    String exp = sc.nextLine();
    exp = exp.replaceAll(" ", "");

    String postfix = infixToPostfix(exp);
    System.out.println();

    System.out.println("Postfix: " + postfix);
    System.out.println();

    int hasil = evaluatePostfix(postfix);
    System.out.println();

    System.out.println("Hasil akhir= " + hasil);

    sc.close();
}
}